tpi_rs/src/middleware/serial.rs

```diff
@@ -0,0 +1,253 @@
+//! Handlers for UART connections to/from nodes
+
+use std::sync::Arc;
+
+use anyhow::Result;
+use bytes::BytesMut;
+use futures::stream::{SplitSink, SplitStream};
+use futures::{SinkExt, StreamExt};
+use tokio::sync::Mutex;
+use tokio::task::JoinHandle;
+use tokio_serial::{DataBits, Parity, SerialPortBuilderExt, SerialStream, StopBits};
+use tokio_util::codec::{BytesCodec, Decoder, Framed};
+
+use super::NodeId;
+
+const OUTPUT_BUF_SIZE: usize = 16 * 1024;
+
+type Sink = SplitSink<Framed<SerialStream, BytesCodec>, BytesMut>;
+type Stream = SplitStream<Framed<SerialStream, BytesCodec>>;
+
+#[derive(Debug)]
+pub struct SerialConnections {
+    handlers: Vec<Handler>,
+}
+
+#[derive(Debug)]
+struct Handler {
+    node: usize,
+    sink: Sink,
+    stream: Arc<Mutex<Stream>>,
+    buf: Arc<Mutex<RingBuffer<OUTPUT_BUF_SIZE>>>,
+    worker: Option<JoinHandle<()>>,
+}
+
+#[derive(Debug)]
+struct RingBuffer<const C: usize> {
+    buf: Vec<u8>,
+    idx: usize,
+    len: usize,
+}
+
+impl SerialConnections {
+    pub fn new() -> Result<Self> {
+        let paths = ["/dev/ttyS2", "/dev/ttyS1", "/dev/ttyS4", "/dev/ttyS5"];
+
+        let mut handlers = vec![];
+
+        for (i, path) in paths.iter().enumerate() {
+            handlers.push(Handler::new(i + 1, path)?);
+        }
+
+        Ok(SerialConnections { handlers })
+    }
+
+    pub fn run(&mut self) {
```

Comment:
* I would like to start/stop to be executed when a node is turned on/off, respectively. 
* There should also be a default option stored inside our `Config` object.  So users have the option to set ' uart reading` to false and therefore disable this feature. ( please be aware this config is intended to be 'read only').

Reply:
Are you sure? Threads/executors should just sit idly consuming no resources while waiting for OS to wake them up.